src/gui/menus/UndoRedoMenu.cpp

```diff
@@ -0,0 +1,159 @@
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "UndoRedoMenu.h"
+
+#include <QApplication>
+#include <QContextMenuEvent>
+#include <QFileInfo>
+#include <QMessageBox>
+#include <QTextStream>
+
+#include "Engine.h"
+#include "Song.h"
+
+#include "embed.h"
+#include "GuiApplication.h"
+#include "MainWindow.h"
+#include "ProjectJournal.h"
+
+namespace lmms::gui
+{
+
+
+UndoRedoMenu::UndoRedoMenu(QWidget *parent) :
+	SideBarWidget(tr("Project History"), embed::getIconPixmap("edit_undo").transformed(QTransform().rotate(90)), parent)
+{
+	m_undoRedoTree = new UndoRedoTreeWidget(contentParent());
+	addContentWidget(m_undoRedoTree);
+	reloadTree();
+	connect(Engine::projectJournal(), &ProjectJournal::checkPointAdded, this, &UndoRedoMenu::reloadTree);
+	connect(Engine::projectJournal(), &ProjectJournal::undoTriggered, this, &UndoRedoMenu::reloadTree);
+	connect(Engine::projectJournal(), &ProjectJournal::redoTriggered, this, &UndoRedoMenu::reloadTree);
+}
+
+
+void UndoRedoMenu::reloadTree()
+{
+	m_undoRedoTree->clear();
+
+	auto& undoHistory = Engine::projectJournal()->undoCheckPoints();
+	auto& redoHistory = Engine::projectJournal()->redoCheckPoints();
+
+	int index = redoHistory.size() - 1;
+	for (auto it = redoHistory.begin(); it != redoHistory.end(); ++it)
+	{
+		m_undoRedoTree->addTopLevelItem(new UndoRedoTreeWidgetItem(m_undoRedoTree, &(*it), index, false));
+		--index;
+	}
+
+	QTreeWidgetItem* divider = new QTreeWidgetItem(m_undoRedoTree);
+	divider->setText(0, tr("--- Current State ---"));
+	m_undoRedoTree->addTopLevelItem(divider);
+
+	index = 0;
+	for (auto it = undoHistory.rbegin(); it != undoHistory.rend(); ++it)
+	{
+		m_undoRedoTree->addTopLevelItem(new UndoRedoTreeWidgetItem(m_undoRedoTree, &(*it), index, true));
+		++index;
+	}
+}
+
+
+UndoRedoTreeWidget::UndoRedoTreeWidget(QWidget* parent):
+	QTreeWidget(parent)
+{
+	setColumnCount(1);
+	headerItem()->setHidden(true);
+	connect(this, &UndoRedoTreeWidget::itemDoubleClicked, this, &UndoRedoTreeWidget::applyUndoRedoEntry);
+}
+
+
+void UndoRedoTreeWidget::contextMenuEvent(QContextMenuEvent * e)
+{
+	if (itemAt(e->pos()) == nullptr) { return; }
+	UndoRedoTreeWidgetItem* entry = dynamic_cast<UndoRedoTreeWidgetItem*>(itemAt(e->pos()));
+	if (entry == nullptr) { return; }
+
+	QMenu contextMenu(this);
+
+	contextMenu.addAction(QIcon(embed::getIconPixmap("edit_undo")), tr(entry->m_isUndo ? "Undo" : "Redo"), [this, entry]{ applyUndoRedoEntry(entry, 0); });
```

Comment:
```suggestion
	contextMenu.addAction(QIcon(embed::getIconPixmap("edit_undo")), entry->m_isUndo ? tr("Undo") : tr("Redo"), [this, entry]{ applyUndoRedoEntry(entry, 0); });
```
I don't think this is a correct way of using `QObject::tr`. It should just be a simple compile-time known string literal.
https://wiki.qt.io/QtInternationalization#tr()_and_encoding